File: Solutions/1161. Maximum Level Sum of a Binary Tree/1161. Maximum Level Sum of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
//DFS
class Solution {
public:
    vector<int> sums;
    void dfs(TreeNode* root,size_t level){
        if(root){
            sums.resize(max(sums.size(),level));
            sums[level-1]+=root->val;
            dfs(root->left,level+1);
            dfs(root->right,level+1);
        }
    }
    int maxLevelSum(TreeNode* root) {
        dfs(root,1);
        return distance(begin(sums),max_element(begin(sums),end(sums)))+1;
    }
};



//BFS , level order traversal
class Solution {
public:
    int maxLevelSum(TreeNode* root) {
        int maxsum=INT_MIN,level=0,maxlevel=0;
        queue<TreeNode*> q;
        q.push(root);
        while(q.size()){
            ++level;
            int sum=0,qsz=q.size();
            while(qsz){
                --qsz;
                auto tmp=q.front();q.pop();
                sum+=tmp->val;
                if(tmp->left)
                    q.push(tmp->left);
                if(tmp->right)
                    q.push(tmp->right);
            }
            if(maxsum<sum){
                maxlevel=level;
                maxsum=sum;
            }
        }
        return maxlevel;
    }
};

File: Solutions/1161. Maximum Level Sum of a Binary Tree/1161. Maximum Level Sum of a Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
//BFS , level order traversal
func maxLevelSum(root *TreeNode) int {
    maxsum,level,maxlevel:=math.MinInt64,0,0
    var q []*TreeNode
    q=append(q,root)
    for len(q)>0{
        level+=1
        sum,qsz:=0,len(q)
        for qsz>0{
            qsz-=1
            tmp:=q[0]
            q=q[1:]
            sum+=tmp.Val
            if tmp.Left!=nil{
                q=append(q,tmp.Left)
            }
            if tmp.Right!=nil{
                q=append(q,tmp.Right)
            }
        }
        if maxsum<sum{
            maxsum,maxlevel=sum,level
        }   
    }
    return maxlevel
}

//DFS
func maxLevelSum(root *TreeNode) int {
    var list []int
    dfs(&list,root,0)
    maxlevel:=0
    for i:=0;i<len(list);i+=1{
        if list[maxlevel]<list[i]{
            maxlevel=i
        }
    }
    return maxlevel+1
}
func dfs(list *[]int,root *TreeNode,level int) {
    if(root==nil){return}
    if(level==len(*list)){
        *list=append(*list,root.Val)
    }else{
        (*list)[level]+=root.Val
    }
    dfs(list,root.Left,level+1)
    dfs(list,root.Right,level+1)
}